src/main/java/com/Elessar/app/server/LogOnHandler.java

```diff
@@ -52,26 +56,39 @@ public void handle(HttpExchange he) throws IOException {
             final LogonRequest logonRequest = LogonRequest.parseFrom(is);
             final LogonResponse.Builder logonResponse = LogonResponse.newBuilder();
             final String userName = logonRequest.getName();
+            final String password = logonRequest.getPassword();
             final String clientURL = new StringBuilder().append("http://")
-                                                               .append(he.getRemoteAddress().getAddress().getHostAddress())
-                                                               .append(":")
-                                                               .append(logonRequest.getPort()).toString();
+                                                        .append(he.getRemoteAddress().getAddress().getHostAddress())
+                                                        .append(":")
+                                                        .append(logonRequest.getPort()).toString();
+
+            final User existingUser = users.getUnchecked(userName);
+            final User newUser = new User(userName, password, null, null, clientURL, true);
 
-            final User prevUser = db.update(new User(userName, logonRequest.getPassword(), null, null, clientURL, true));
+            if (existingUser.getName() == null || !password.equals(existingUser.getPassword())) {
+                // remove dummy entry as this is an invalid request
+                users.invalidate(userName);
```

Comment:
small improvement can be made here:
if we are here due to password mismatches then we probably should NOT invalidate cache as the user exists.